transport/ProbeServer.cpp

```diff
@@ -0,0 +1,369 @@
+{
+    uint64_t timestamp = utils::Time::getAbsoluteTime();
+    const void* data = packet->get();
+
+    auto* stunMessage = ice::StunMessage::fromPtr(data);
+
+    if (stunMessage && stunMessage->isValid() && stunMessage->header.isRequest() &&
+        stunMessage->isAuthentic(_credentials.second))
+    {
+        ice::StunMessage response;
+        response.header.transactionId = stunMessage->header.transactionId;
+        response.header.setMethod(ice::StunHeader::BindingResponse);
+        response.add(ice::StunXorMappedAddress(destination, response.header));
+        response.addMessageIntegrity(_credentials.second);
+        response.addFingerprint();
+
+        endpoint.sendStunTo(destination, response.header.transactionId.get(), &response, response.size(), timestamp);
+    }
+    else
+    {
+        endpoint.stop(this);
+    }
+}
+
+void ProbeServer::addCandidate(const ice::IceCandidate& candidate)
+{
+    if (!utils::contains(_iceCandidates, [candidate](const ice::IceCandidate& x) {
+            return x.address == candidate.address && x.baseAddress == candidate.baseAddress &&
+                x.transportType == candidate.transportType;
+        }))
+    {
+        _iceCandidates.push_back(candidate);
+    }
+
+    std::sort(_iceCandidates.begin(), _iceCandidates.end(), [](const ice::IceCandidate& a, const ice::IceCandidate& b) {
+        return a.priority > b.priority;
+    });
+}
+
+void ProbeServer::run()
+{
+    concurrency::setThreadName("ProbeServerMaintenance");
+
+    while (_maintenanceThreadIsRunning)
+    {
+        while (!_queue.empty())
+        {
+            ProbeTcpConnection connection;
+            if (_queue.pop(connection))
+            {
+                _tcpConnections.emplace_back(connection);
+            }
+        }
+
+        auto now = utils::Time::getAbsoluteTime();
+
+        for (auto it = _tcpConnections.begin(); it != _tcpConnections.end(); ++it)
+        {
+            if (utils::Time::diffGT(it->timestamp,
+                    now,
+                    _iceConfig.probeConnectionExpirationTimeout * utils::Time::sec)) // TODO use config value
+            {
+                it->endpoint->stop(this);
+                _tcpConnections.erase(it);
+            }
+        }
+
+        utils::Time::nanoSleep(100 * utils::Time::ms);
+    }
+}
+
+void ProbeServer::stop()
+{
+    _maintenanceThreadIsRunning = false;
+
+    while (!_queue.empty())
+    {
+        ProbeTcpConnection connection;
+        if (_queue.pop(connection))
+        {
+            _tcpConnections.emplace_back(connection);
+        }
+    }
+
+    for (auto it = _tcpConnections.begin(); it != _tcpConnections.end(); ++it)
+    {
+        it->endpoint->stop(this);
+        _tcpConnections.erase(it);
+    }
+}
+
+// ICE
+const std::pair<std::string, std::string>& ProbeServer::getCredentials() const
+{
+    return _credentials;
+}
+
+const std::vector<ice::IceCandidate> ProbeServer::getCandidates()
```

Comment:
We should consider return a const reference instead and avoid copies of vector each time we need the candidates

Reply:
Also can you ensure this is thread safe? I mean we are not making a copy (or returning a reference) when the vector is being changed by other thread?